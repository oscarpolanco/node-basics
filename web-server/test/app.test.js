const superTest = require('supertest');
const request = require('request');
const app = require('../src/app');
jest.mock('request');

beforeEach(() => jest.clearAllMocks());

test('Should return a error message if a location is not provide', async () => {
    const response = await superTest(app)
        .get('/weather?address=')
        .send()
        .expect(200);
    
        expect(response.body.error).toEqual('You must provide an address!');
});

test('Should return the unable of connect error message when the location service was unavailable', async () => {
    request.mockImplementation((config, cb) => cb({ error: 'error' }, {}));


    const response = await superTest(app)
        .get('/weather?address=boston')
        .send()
        .expect(200);

    expect(response.body.error).toEqual('Unable to connect to location service!');
});

test('Should return the unable of connect error message when the weather service was unavailable', async () => {
    request.mockImplementationOnce((config, cb) => cb(undefined, {
            body: {
                features: [{
                    center: [-71.058291, 42.360253],
                    place_name: 'Boston, Massachusetts, United States'
                }]
            }})
    )
    .mockImplementationOnce((config, cb) => {
        return cb({ error: 'error' }, {});
    });

    const response = await superTest(app)
        .get('/weather?address=boston')
        .send()
        .expect(200);

    expect(response.body.error).toEqual('Unable to connect to weather service!');
});

test('Should return the unable to find location for a no existing place for the location service', async () => {
    request.mockImplementation((config, cb) => {
        return cb({}, { features: [] })
    });

    const response = await superTest(app)
        .get('/weather?address=foo')
        .send()
        .expect(200);

    expect(response.body.error).toEqual('Unable to connect to location service!');
});

test('Should return the unable to find location for a no existing place for the weather service', async () => {
    request.mockImplementationOnce((config, cb) => {
        return cb(undefined, { body: {features: [{
            center: [-71.058291, 42.360253],
            place_name: 'Boston, Massachusetts, United States'
        }]}});
    })
    .mockImplementationOnce((config, cb) => {
        return cb(undefined, { body: {
            error: 'error'
        }});
    });

    const response = await superTest(app)
        .get('/weather?address=boston')
        .send()
        .expect(200);

    expect(response.body.error).toEqual('Unable to find location');
});

test('Should return the weather information from a valid location', async () => {
    request.mockImplementationOnce((config, cb) => cb(undefined, {
        body: {
            features: [{
                center: [-71.058291, 42.360253],
                place_name: 'Boston, Massachusetts, United States'
            }]
        }})
    )
    .mockImplementationOnce((config, cb) => cb(undefined, {
        body: {
            current: {
                weather_descriptions: [ 'Clear' ],
                temperature: 46,
                feelslike: 43,
                humidity: 49,
            }
        }})
    );

    const response = await superTest(app)
        .get('/weather?address=boston')
        .send()
        .expect(200);

    expect(response.body).toMatchObject({
        forecast: 'Clear. Its is currently 46 degrees out. Its feels like 43 degrees out and the humidity is 49%',
        location: 'Boston, Massachusetts, United States',
        address: 'boston'
    });
});

test('Should response to the about page', async () => {
    const response = await superTest(app)
        .get('/about')
        .send()
        .expect(200);

    expect(response.text).not.toBeNull();
    expect(response.text).toMatchSnapshot();
});

test('Should response to the help page', async () => {
    const response = await superTest(app)
        .get('/help')
        .send()
        .expect(200)

    expect(response.text).not.toBeNull();
    expect(response.text).toMatchSnapshot();
});

test('Should return the 404 page for a not existing route', async () => {
    const response = await superTest(app)
        .get('/foo')
        .send()
        .expect(200);

    expect(response.text).not.toBeNull();
    expect(response.text).toMatchSnapshot();
});
