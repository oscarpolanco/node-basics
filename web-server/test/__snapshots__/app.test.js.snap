// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should response to the about page 1`] = `
"<!DOCTYPE html>
<html lang="en">
    <head>
        <title>About</title>
        <link rel="icon" href="/img/weather.png">
        <link rel="stylesheet" href="/css/styles.css">
    </head>
    <body>
        <div class="main-content">
            <header>
                <h1>About Me</h1>
            
                <a href="/">Weather</a>
                <a href="/about">About</a>
                <a href="/help">Help</a>
            </header>
            <p>This site was created by me. It use data from mapbox.com and weatherstack.com!</p>
            <img class="portrait" src="/img/robot.png" />
        </div>
        <footer>
            <p>Created by Testing</p>
        </footer>
    </body>
</html>
"
`;

exports[`Should response to the help page 1`] = `
"<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Help</title>
        <link rel="icon" href="/img/weather.png">
        <link rel="stylesheet" href="/css/styles.css">
    </head>
    <body>
        <div class="main-content">
            <header>
                <h1>Help</h1>
            
                <a href="/">Weather</a>
                <a href="/about">About</a>
                <a href="/help">Help</a>
            </header>
            <p>This is some helpful message</p>
        </div>
        <footer>
            <p>Created by Testing</p>
        </footer>
    </body>
</html>
"
`;

exports[`Should return the 404 page for a not existing route 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <title>404</title>
        <link rel="icon" href="/img/weather.png">
        <link rel="stylesheet" href="/css/styles.css">
    </head>
    <body>
        <div class="main-content">
            <header>
                <h1>404</h1>
            
                <a href="/">Weather</a>
                <a href="/about">About</a>
                <a href="/help">Help</a>
            </header>
            <p>Page not found</p>
        </div>
        <footer>
            <p>Created by Testing</p>
        </footer>
    </body>
</html>
"
`;
